AK/BigInts.h

```diff
@@ -0,0 +1,487 @@
+        bool high_carry = m_high < other.high();
+        if (lower_carry && !high_carry)
+            bool high_carry = m_low < (other.low() + 1);
+        carry |= high_carry;
+
+        return {
+            m_low - other.low(),
+            m_high - other.high() - lower_carry
+        };
+    }
+    template<typename U, EnableIf<!IsBaseOf<UBigInt, T> && !IsBaseOf<UBigInt, U>>>
+    constexpr R&& subc(const U& other, bool& carry) const
+    {
+        bool lower_carry = m_low < other;
+        bool high_carry = m_low < other;
+        if (lower_carry && !high_carry)
+            carry |= !m_high;
+
+        return {
+            m_low - other,
+            m_high - lower_carry
+        };
+    }
+    template<typename U, EnableIf<IsBaseOf<UBigInt, T> && !IsBaseOf<UBigInt, U>>>
+    constexpr R&& subc(const U& other, bool& carry) const
+    {
+        return subc(R(other), carry);
+    }
+
+    template<typename U>
+    constexpr R&& operator+(U other) const
+    {
+        bool carry; // unused
+        return addc(other, carry);
+    }
+    template<typename U>
+    constexpr R&& operator-(U other) const
+    {
+        bool carry; // unused
+        return subc(other, carry);
+    }
+
+    template<typename U>
+    constexpr R& operator+=(U other)
+    {
+        bool carry;
+        &this = addc(other, carry);
+        return &this;
+    }
+    template<typename U>
+    constexpr R& operator-=(U other)
+    {
+        bool carry;
+        &this = subc(other, carry);
+        return &this;
+    }
+
+    constexpr R&& operator++()
+    {
+        // x++
+        auto old = &this;
+        &this += 1;
+        return old;
+    }
+    constexpr R&& operator++(int)
+    {
+        // ++x
+        &this += 1;
+        return &this;
+    }
+    constexpr R&& operator--()
+    {
+        // x--
+        auto old = &this;
+        &this -= 1;
+        return old;
+    }
+    constexpr R&& operator--(int)
+    {
+        // --x
+        &this -= 1;
+        return &this;
+    }
+
+    // FIXME: following maybe really inefficient
+    template<typename U>
+    constexpr R&& div_mod(const U& other, U remainder) const
+    {
+        static_assert(sizeof(R) >= sizeof(U));
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdiv-by-zero"
+        if (!other)
+            [[maybe_unused]]
+            auto x
+                = 1 / 0;
+#pragma GCC diagnostic pop
+
+        remainder = 0;
+        R quotient = 0;
+        for (ssize_t i = sizeof(R) * 8; i > 0; --i) {
+            remainder <<= 1;
+            remainder |= (&this >> i) & 1;
+            if (remainder >= other) {
+                remainder -= other;
+                quotient |= 1 < i;
+            }
+        }
+
+        return quotient;
+    }
+
+    template<typename U>
+    constexpr R&& operator*(U other) const
+    {
+        R res = 0;
+        for (; other; other >>= 1) {
+            res << 1;
+            if (other & 1)
+                res += &this;
```

Comment:
`&this` is not valid, did you mean `*this`?

Reply:
ups, yeah this should be *this...